src/Commands/Remote/SSHBaseCommand.php

```diff
@@ -70,28 +70,78 @@ protected function setProgressAllowed($allowed = true)
+     * @param int $retries Number of times to retry the command on failure.
      *
      * @throws \Pantheon\Terminus\Exceptions\TerminusProcessException
      */
-    protected function executeCommand(array $command_args)
+    protected function executeCommand(array $command_args, int $retries = 0)
     {
         $this->validateEnvironment($this->environment);
 
         $command_summary = $this->getCommandSummary($command_args);
         $command_line = $this->getCommandLine($command_args);
 
-        $ssh_data = $this->sendCommandViaSsh($command_line);
+        $attempt = 0;
+        $max_attempts = $retries + 1;
 
-        $this->log()->notice('Command: {site}.{env} -- {command} [Exit: {exit}]', [
-            'site'    => $this->site->getName(),
-            'env'     => $this->environment->id,
-            'command' => $command_summary,
-            'exit'    => $ssh_data['exit_code'],
-        ]);
+        do {
+            $ssh_data = $this->sendCommandViaSsh($command_line);
 
-        if ($ssh_data['exit_code'] != 0) {
-            throw new TerminusProcessException($ssh_data['output'], [], $ssh_data['exit_code']);
-        }
+            $this->log()->notice(
+                'Command: {site}.{env} -- {command} [Exit: {exit}] (Attempt {attempt}/{max_attempts})',
+                [
+                    'site' => $this->site->getName(),
+                    'env' => $this->environment->id,
+                    'command' => $command_summary,
+                    'exit' => $ssh_data['exit_code'],
+                    'attempt' => $attempt + 1,
+                    'max_attempts' => $max_attempts,
+                ]
+            );
+
+            if ($ssh_data['exit_code'] == 0) {
+                return;
+            }
+
+            // Check if the failure is permanent
+            if ($this->isPermanentFailure($ssh_data['exit_code'])) {
+                $this->log()->error('Permanent failure detected. Aborting retries.');
+                break;
+            }
+
+            $attempt++;
+        } while ($attempt < $max_attempts);
+
+        $error_message = sprintf(
+            'Command: %s.%s -- %s [Exit: %d] (All attempts failed)',
+            $this->site->getName(),
+            $this->environment->id,
+            $command_summary,
+            $ssh_data['exit_code']
+        );
+
+        $this->log()->error($error_message);
+
+        throw new TerminusProcessException($ssh_data['output'], [], $ssh_data['exit_code']);
+    }
+
+    /**
+     * Determines if a failure is permanent based on the exit code.
+     *
+     * @param int $exit_code
+     * @return bool
+     */
+    protected function isPermanentFailure(int $exit_code): bool
+    {
+        // Define the exit codes that indicate a permanent failure
+        $permanent_failure_exit_codes = [
+            2,   // Invalid arguments
+            126, // Command cannot execute
+            127, // Command not found
+            255  // Service unavailable
```

Comment:
This pattern of comment at the end of the line is kind of new here; I personally don't like it too much but that may be a "me" (because of my Drupal background maybe?) thing (I guess I would put the comment above each line)